### ISSUE TITLE
Support for CSS Custom Variables

### ISSUE THREAD

**Issue by mjgchase** [Automattic/juice#312]
Would be nice to support and convert css custom variables 
https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_variables

**Comment by NC-piercej** [comment]
Since email clients have generally poor support for CSS variables, it would be nice if Juice would inline them using `getComputedStyle`.

**Comment by benaltair** [comment]
I agree - passing on the variable doesn't work, so computing it and inlining the computed value would be stellar. I have tested variables and it didn't work as expected, with missing styles on the email client side.

**Comment by huksley** [comment]
Any ETA on implementing this? Very important as I use components and have colors, borders, etc defined once and used everywhere, including email templates.

**Comment by jrit** [comment]
@huksley would love to take a PR from you if you have time to do it!

**Comment by shaunpersad** [comment]
For anyone looking for some rudimentary support for replacing CSS variables after juice is done, try this:
```typescript
function replaceCssVariables(str: string): string {
  const match = str.match(regex);
  if (!match) {
    return str;
  }
  const { index } = match;
  const [found, key, value] = match;
  const alreadyProcessed = str.slice(0, index! + found.length);
  const remaining = str.slice(index! + found.length).replace(new RegExp(`var\\(--${key}\\)`, 'g'), value);
  return `${alreadyProcessed}${replaceCssVariables(remaining)}`;
}

  const inlined = juice.inlineContent(html, css);
  const replaced = replaceCssVariables(inlined);
```

It works excellent for my use cases, but YMMV.

**Comment by shaunpersad** [comment]
Here's a solution for replacing the CSS variables after juice is done, using cheerio:
```typescript
const variableUsageRegex = /:\s*var\((--[a-zA-Z0-9-_]+)\);*/g;
const variableDefRegex = /(--[a-zA-Z0-9-_]+):\s*([^;]+);/g;

/**
 * Returns an array of variable usages.
 * Usages look like var(--name)
 */
function findCssVariableUsages(str = ''): string[] {
  const matches = Array.from(str.matchAll(variableUsageRegex));
  return matches.map((match) => match[1]);
}

/**
 * Returns a map of variable defs.
 * Defs look like --name: foo
 */
function findCssVariableDefs(str = ''): Map<string, string> {
  const matches = Array.from(str.matchAll(variableDefRegex));
  return matches.reduce((defs, [_, key, value]) => {
    defs.set(key, value);
    return defs;
  }, new Map<string, string>());
}

/**
 * Replaces CSS variables by iterating through each element and for each one,
 * if variable usages are found in its style,
 * walk up its ancestors to find the variable definitions
 * and replace the usages with the definitions.
 */
function replaceCssVariables(html = ''): string {
  const $ = cheerio.load(html, null, false);
  $('*').each((index, el) => {
    let $el = $(el);
    let style = $el.attr('style') || '';
    let variables = findCssVariableUsages(style);
    while (variables.length && ($el = $el.parent())) {
      const defs = findCssVariableDefs($el.attr('style'));
      variables.forEach((variable) => {
        if (defs.has(variable)) {
          style = style.replace(new RegExp(`var\\(${variable}\\)`, 'g'), defs.get(variable) || '');
          variables = variables.filter((v) => v !== variable);
        }
      });
    }
    $(el).attr('style', style);
  });
  return cheerio.html($.root());
}
```

**Comment by pgodman** [comment]
> Here's a solution for replacing the CSS variables after juice is done, using cheerio:
> 
> ```ts
> const variableDefRegex = /^(--[a-zA-Z0-9-_]+)$/;
> const variableUsageRegex = /var\((--[a-zA-Z0-9-_]+)(?:\)|,\s*(.*)\))/;
> /**
>  * Resolves any variable usages found in a style value to their definitions.
>  */
> export function resolveCSSVariables(defs: Map<string, string>, styleValue = ''): string {
>   let match;
>   while (match = styleValue.match(variableUsageRegex)) {
>     const [found, variableName, fallback] = match;
>     const { index } = match;
>     const before = styleValue.slice(0, index);
>     const after = styleValue.slice(index! + found.length);
>     const replacement = defs.has(variableName)  ? defs.get(variableName) : fallback;
>     styleValue = `${before}${resolveCSSVariables(defs, replacement || '')}${after}`;
>   }
>   return styleValue;
> }
> 
> /**
>  * Walks the tree depth-first,
>  * collecting variable definitions and then resolving any usages found.
>  */
> export function replaceCSSVariables($: CheerioAPI, $el : Cheerio<any>, defs = new Map<string, string>()) {
>   const styles = $el.css();
>   if (styles) {
>     /**
>      * Collect the defs first.
>      */
>     Object.entries(styles).forEach(([key, value]) => {
>       if (variableDefRegex.test(key)) {
>         defs.set(key, value);
>       }
>     });
>     /**
>      * Resolve any usages.
>      * Done separately from above in case any defs and usages are found on the same element.
>      */
>     Object.entries(styles).forEach(([key, value]) => {
>       styles[key] = resolveCSSVariables(defs, value);
>     });
>     $el.css(styles);
>   }
>   $el.children().each(
>     (index, el) => replaceCSSVariables($, $(el), new Map<string, string>(defs)),
>   );
> }
> 
> export default function inlineWithVariablesReplaced(html: string, css: string): string {
>   const inlined = juice.inlineContent(html, css, { inlinePseudoElements: true });
>   const $ = cheerio.load(inlined, null, false);
>   const $root = $.root();
>   replaceCSSVariables($, $root);
>   return cheerio.html($root);
> }
> ```

This is great. Am I correct in understanding, though, that if the variable is defined in various classes that apply to an element, this will just use the most-recently-encountered variable definition that matches, rather than the definition from the highest-priority class?
